In [1]:
import pandas as pd

link = r"C:\Users\ajayp\OneDrive\Documents\All dataset at one place\Cutomer_churn_analysis.xlsx"

df = pd.read_excel(link)

In [2]:
df.head()

,Customer ID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,Unnamed: 8,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [3]:
df.shape

(7043, 21)

In [4]:
df.columns

Index(['Customer ID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'Unnamed: 8',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Customer ID       7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   Unnamed: 8        7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [6]:
df.isnull().sum()

Customer ID         0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
Unnamed: 8          0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [7]:
#Checking duplicate rows
df["Customer ID"].duplicated().sum()

np.int64(0)

In [8]:
df["Unnamed: 8"].value_counts()

Unnamed: 8
Fiber optic    3096
DSL            2421
No             1526
Name: count, dtype: int64

In [9]:
df[["MultipleLines", "Unnamed: 8", "OnlineSecurity"]].head(10)

,MultipleLines,Unnamed: 8,OnlineSecurity
0,No phone service,DSL,No
1,No,DSL,Yes
2,No,DSL,Yes
3,No phone service,DSL,Yes
4,No,Fiber optic,No
5,Yes,Fiber optic,No
6,Yes,Fiber optic,No
7,No phone service,DSL,Yes
8,Yes,Fiber optic,No
9,No,DSL,Yes


In [10]:
# renaming the blanck column
df.rename(columns = {"Unnamed: 8" : "InternetService"}, inplace=True)
df.columns

Index(['Customer ID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [11]:
#Exposing the Hidden Missing Values

df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')

df.isnull().sum()

Customer ID          0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

In [12]:
#look specifically at those 11 rows

df[df["TotalCharges"].isnull()]

,Customer ID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
488,4472-LVYGI,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,Yes,No,Two year,Yes,Bank transfer (automatic),52.55,NaN,No
753,3115-CZMZD,Male,0,No,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.25,NaN,No
936,5709-LVOEQ,Female,0,Yes,Yes,0,Yes,No,DSL,Yes,...,Yes,No,Yes,Yes,Two year,No,Mailed check,80.85,NaN,No
1082,4367-NUYAO,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.75,NaN,No
1340,1371-DWPAZ,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,Yes,No,Two year,No,Credit card (automatic),56.05,NaN,No
3331,7644-OMVMY,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,19.85,NaN,No
3826,3213-VVOLG,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.35,NaN,No
4380,2520-SGTTA,Female,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.00,NaN,No
5218,2923-ARZLG,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,One year,Yes,Mailed check,19.70,NaN,No
6670,4075-WKNIU,Female,0,Yes,Yes,0,Yes,Yes,DSL,No,...,Yes,Yes,Yes,No,Two year,No,Mailed check,73.35,NaN,No


This confirms that the above customers have just signed up, so their time with the company (tenure) is exactly **0 months**, and they haven't paid their first bill yet.


During our data inspection, we found 11 missing values in the TotalCharges column. By investigating further, we discovered that all 11 of these customers have a tenure of 0. This means they are brand new customers who have just signed up and haven't been billed yet.

Since our main goal is to understand customer churn (why people leave), brand new customers don't give us any historical data to learn from. Furthermore, 11 rows represent a tiny fraction (about 0.15%) of our 7,043 total records. Because of this, the most logical step is to drop these 11 rows from our analysis so they don't skew our results.

In [13]:
df.dropna(inplace=True)
df.shape

(7032, 21)

In [14]:
# Checking the number of Churns

df["Churn"].value_counts()

Churn
No     5163
Yes    1869
Name: count, dtype: int64

In [15]:
df["Churn"].value_counts(normalize =True)*100

Churn
No     73.421502
Yes    26.578498
Name: proportion, dtype: float64

Calculating the Churn Rate

While raw counts are helpful for understanding volume, business stakeholders track success using percentages. The **Churn Rate** represents the proportion of our total customer base that has left.


In [16]:
pd.crosstab(df["Contract"],df["Churn"])

Churn,No,Yes
Contract,,
Month-to-month,2220,1655
One year,1306,166
Two year,1637,48


This confirms that the Month-to-Month contracts have a higher likelihood of churning compared to those on 1-year or 2-year commitments.

In [17]:
# Checking avg tenure of churning customers
df.groupby("Churn")["tenure"].mean()

Churn
No     37.650010
Yes    17.979133
Name: tenure, dtype: float64

We found a significant difference in customer lifespan. Customers who stay have an average tenure of roughly 38 months, whereas customers who churn leave after an average of just 18 months. This indicates that the first year and a half is the highest-risk period for customer attrition.

In [18]:
# Analyzing the Impact of Price
df.groupby("Churn")["MonthlyCharges"].mean()

Churn
No     61.307408
Yes    74.441332
Name: MonthlyCharges, dtype: float64

We investigated whether price impacts customer retention. The data shows that customers who churn have a significantly higher average monthly bill compared to those who stay. This suggests that price sensitivity is a major factor in attrition.

In [23]:
# Now saving this cleaned file in excel 

df.to_csv(r"C:\Users\ajayp\OneDrive\Documents\All dataset at one place\Cutomer_churn_analysis_Cleaned_.csv", index=False)

Data Cleaning Actions Taken-

Identified and Renamed Columns: Renamed the ambiguous Unnamed: 8 column to InternetService for accurate analysis.

Corrected Data Types: Converted TotalCharges from a text object to a numerical float, which successfully exposed hidden missing values stored as blank spaces.

Handled Missing Data: Discovered 11 missing values in TotalCharges linked to brand-new customers (0 months tenure) and dropped these rows, as they provided no historical churn data.

Verified Data Integrity: Confirmed the dataset contained zero perfectly duplicated rows, ensuring accurate counting for our metrics.

Key Business Insights Discovered

Baseline Churn Rate: The company is currently experiencing an overall customer churn rate of 26.58%.

Contract Vulnerability: Customers on Month-to-month contracts make up the vast majority of churned accounts (1,655 out of 1,869), proving that lacking a long-term commitment is a primary risk factor.

Critical Retention Window: Retained customers stay for an average of nearly 38 months, while churning customers leave after an average of just 18 months, highlighting a critical period for proactive retention efforts.

Price Sensitivity: Customers who churned paid a higher average monthly bill ($74.44) compared to those who stayed ($61.31), suggesting price sensitivity or perceived lack of value among higher-paying tiers.